README.md
# GitHub

[![Build Status](https://img.shields.io/travis/WorkflowCI/github.svg)](https://travis-ci.org/WorkflowCI/github)
[![Coverage Status](https://coveralls.io/repos/github/WorkflowCI/github/badge.svg)](https://coveralls.io/github/WorkflowCI/github)
[![Latest Version](https://img.shields.io/hexpm/v/github.svg)](https://hex.pm/packages/github)

The simplest Elixir client for GitHub [REST API v3](https://developer.github.com/v3/).

## Contents

* [Usage](#usage)
  * [With App JWT token](#with-app-jwt-token)
  * [With OAuth access token](#with-oauth-access-token)
  * [Pagination](#pagination)
* [API Resouces](#api-resources)
* [Installation](#installation)
* [Configuration](#configuration)
* [Testing](#testing)

## Usage

### With App JWT token

To authenticate with a GitHub App and make an API call by using a [JWT token](https://developer.github.com/apps/building-github-apps/authenticating-with-github-apps/):

```elixir
iex> jwt_token = Github.Client.generate_jwt_token(app_id: 12345, private_key_filepath: "app.pem")

iex> github_client = %Github.Client{jwt_token: "jwt_token"}

iex> github_client |> Github.Apps.Installations.find!(67890)
%Github.Client.Response{
  status: 200,
  headers: [...],
  body: %{"id" => 12345, ...},
  ...
}
```

### With OAuth access token

To get an access token, redirect a user to an authorize url:

```elixir
def github(conn, _params) do
  authorize_url = Github.Oauth2.Client.authorize_url!(
    config: [client_id: "client_id", client_secret: "client_secret"],
    scope: "user:email"
  )
  conn |> redirect(external: authorize_url)
end
```

If a user authorized a log in then GitHub will hit an application callback endpoint with a `code`:

```elixir
def github_callback(conn, %{"code" => code}) do
  access_token = Github.Oauth2.Client.access_token!(
    config: [client_id: "client_id", client_secret: "client_secret"],
    params: [code: code]
  )

  github_client = %Github.Client{access_token: access_token}

  %{"login" => login} = github_client |> Github.Users.find!() |> Map.fetch!(:body)
  [%{"email" => email} | _] = github_client |> Github.Users.Emails.list!() |> Map.fetch!(:body)
  ...
end
```

### Pagination

For pagination, the client has such functions as [Github.Client.fetch_more!](https://hexdocs.pm/github/Github.Client.html#fetch_more!/1) and [Github.Client.fetch_all!](https://hexdocs.pm/github/Github.Client.html#fetch_all!/1). Here is an example:

```elixir
iex> github_client = %Github.Client{access_token: "access_token"}

iex> first_page = github_client |> Github.Users.Emails.list!(page: 1, per_page: 1)
%Github.Client.Response{
  status: 200,
  headers: [...],
  body: [...],
  next_page: 2,
  last_page: 3,
  next_url: "https://api.github.com/user/emails?page=2&per_page=1",
  last_url: "https://api.github.com/user/emails?page=3&per_page=1",
  ...
}

iex> Github.Client.fetch_more!(first_page)
%Github.Client.Response{
  status: 200,
  headers: [...],
  body: [...],
  next_page: 3,
  last_page: 3,
  next_url: "https://api.github.com/user/emails?page=3&per_page=1",
  last_url: "https://api.github.com/user/emails?page=3&per_page=1",
  ...
}
```

Please visit [hexdocs.pm/github](https://hexdocs.pm/github/api-reference.html) to find more code examples.

## API Resources

* [Apps](https://hexdocs.pm/github/Github.Apps.html)
  * [Installations](https://hexdocs.pm/github/Github.Apps.Installations.html)
* Git
  * [References](https://hexdocs.pm/github/Github.Git.Refs.html)
* [Issues](https://hexdocs.pm/github/Github.Issues.html)
  * [Comments](https://hexdocs.pm/github/Github.Issues.Comments.html)
  * [Labels](https://hexdocs.pm/github/Github.Issues.Labels.html)
* [Organizations](https://hexdocs.pm/github/Github.Orgs.html)
* [Pull Requests](https://hexdocs.pm/github/Github.Pulls.html)
* [Repositories](https://hexdocs.pm/github/Github.Repos.html)
  * [Commits](https://hexdocs.pm/github/Github.Repos.Commits.html)
  * [Contents](https://hexdocs.pm/github/Github.Repos.Contents.html)
  * [Statuses](https://hexdocs.pm/github/Github.Repos.Statuses.html)
* Search
  * [Issues](https://hexdocs.pm/github/Github.Search.Issues.html)
* [Users](https://hexdocs.pm/github/Github.Users.html)
  * [Emails](https://hexdocs.pm/github/Github.Users.Emails.html)

## Installation

The package can be installed by adding `github` to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [
    {:github, "~> 0.12.0-rc1"}
  ]
end
```

## Configuration

It uses `jason` as a default JSON library. Change the config to use any other JSON library, for example:

```elixir
config Github.Client, json_library: Poison
```

## Testing

```
mix deps.get
mix test
```


CHANGELOG.md
# Changelog

The following are lists of the notable changes included with each release.
This is intended to help keep people informed about notable changes between
versions, as well as provide a rough history. Each item is prefixed with
one of the following labels: `Added`, `Changed`, `Deprecated`,
`Removed`, `Fixed`, `Security`. We also use [Semantic Versioning](http://semver.org)
to manage the versions of this gem so
that you can set version constraints properly.

#### [Unreleased](https://github.com/WorkflowCI/github/compare/v0.13.0-rc1...HEAD)

* WIP

#### [v0.13.0-rc1](https://github.com/WorkflowCI/github/compare/v0.13.0-rc1...v0.12.0-rc1) – 2018-12-24

* `Added`: [Github.Search.Issues](https://hexdocs.pm/github/Github.Search.Issues.html).
* `Added`: [Github.Issues](https://hexdocs.pm/github/Github.Issues.html).
* `Added`: [Github.Issues.Comments](https://hexdocs.pm/github/Github.Issues.Comments.html).
* `Added`: [Github.Issues.Labels](https://hexdocs.pm/github/Github.Issues.Labels.html).

#### [v0.12.0-rc1](https://github.com/WorkflowCI/github/compare/v0.12.0-rc1...v0.11.0) – 2018-12-22

* `Changed`: Switch from `poison` to `jason`, make it optional.
* `Changed`: `Joken` dependency version from 1.x to 2.x.
* `Changed`: Run tests asynchronously.
* `Added`: Enforce `mix format`.

#### [v0.11.0](https://github.com/WorkflowCI/github/compare/v0.10.0...v0.11.0) – 2018-12-22

* `Added`: [Github.Oauth2.Client](https://hexdocs.pm/github/Github.Oauth2.Client.html).
* `Added`: [Github.Users](https://hexdocs.pm/github/Github.Users.html).

#### [v0.10.0](https://github.com/WorkflowCI/github/compare/v0.9.0...v0.10.0) – 2018-09-04

* `Added`: [Github.Apps](https://hexdocs.pm/github/Github.Apps.html).

#### [v0.9.0](https://github.com/WorkflowCI/github/compare/v0.8.0...v0.9.0) – 2018-09-02

* `Added`: [Github.Pulls](https://hexdocs.pm/github/Github.Pulls.html).
* `Added`: [Github.Apps.Installations.find!](https://hexdocs.pm/github/Github.Apps.Installations.html#find!/2).
* `Added`: [Github.Client.generate_jwt_token](https://hexdocs.pm/github/Github.Client.html#generate_jwt_token/1).

#### [v0.8.0](https://github.com/WorkflowCI/github/compare/v0.7.0...v0.8.0) – 2018-09-02

* `Added`: [Github.Repos.Commits](https://hexdocs.pm/github/Github.Repos.Commits.html).
* `Added`: [Github.Repos.find!](https://hexdocs.pm/github/Github.Repos.html#find!/2).

#### [v0.7.0](https://github.com/WorkflowCI/github/compare/v0.6.0...v0.7.0) – 2018-08-31

* `Added`: [Github.Repos.Contents](https://hexdocs.pm/github/Github.Repos.Contents.html).
* `Added`: [Github.Git.Refs](https://hexdocs.pm/github/Github.Git.Refs.html).

#### [v0.6.0](https://github.com/WorkflowCI/github/compare/v0.5.0...v0.6.0) – 2018-08-31

* `Added`: [Github.Client.fetch_all!](https://hexdocs.pm/github/Github.Client.html#fetch_all!/1).
* `Added`: [Github.Client.fetch_more!](https://hexdocs.pm/github/Github.Client.html#fetch_more!/1).
* `Added`: more code examples.

#### [v0.5.0](https://github.com/WorkflowCI/github/compare/v0.4.0...v0.5.0) – 2018-08-31

* `Added`: [Github.Apps.Installations](https://hexdocs.pm/github/Github.Apps.Installations.html).
* `Added`: [Github.Client](https://hexdocs.pm/github/Github.Client.html).
* `Added`: pagination.

#### [v0.4.0](https://github.com/WorkflowCI/github/compare/v0.3.0...v0.4.0) – 2018-08-31

* `Added`: [Github.Orgs.find!](https://hexdocs.pm/github/Github.Orgs.html#find!/2).
* `Added`: [Github.Client.Response](https://hexdocs.pm/github/Github.Client.Response.html).

#### [v0.3.0](https://github.com/WorkflowCI/github/compare/v0.2.0...v0.3.0) – 2018-08-30

* `Added`: [Github.Users.Emails](https://hexdocs.pm/github/Github.Users.Emails.html).

#### [v0.2.0](https://github.com/WorkflowCI/github/compare/v0.1.0...v0.2.0) – 2018-08-30

* `Added`: [Github.Orgs](https://hexdocs.pm/github/Github.Orgs.html).

#### [v0.1.0](https://github.com/WorkflowCI/github/commit/456368e9fafc3416787ff1275c281f6dad236280) – 2018-08-30

* `Added`: initial functional version with [Github.Repos.Statuses](https://hexdocs.pm/github/Github.Repos.Statuses.html).


lib/github/client.ex
defmodule Github.Client do
  @moduledoc """
    Struct which represents GitHub client
  """

  @link_url_regex ~r/<(?<url>[^>]+)/
  @link_page_regex ~r/\bpage=(?<page>\d+)/
  @generate_jwt_token_default_options %{
    expire_in_minutes: 10
  }

  defstruct [:access_token, :jwt_token]

  def to_json!(map) do
    json_library().encode!(map)
  end

  def from_json!(map) do
    json_library().decode!(map)
  end

  def post!(url, headers, body \\ "") do
    request = %{url: url, body: body, headers: headers}
    response = HTTPoison.post!(url, body, headers)

    %Github.Client.Response{
      request: request,
      body: from_json!(response.body),
      headers: response.headers,
      status: response.status_code
    }
  end

  def get!(url, headers) do
    request = %{url: url, headers: headers}
    response = HTTPoison.get!(url, headers)
    next_link_header = next_link_header(response)
    last_link_header = last_link_header(response)

    %Github.Client.Response{
      request: request,
      body: from_json!(response.body),
      headers: response.headers,
      status: response.status_code,
      next_url: link_url(next_link_header),
      last_url: link_url(last_link_header),
      next_page: link_page(next_link_header),
      last_page: link_page(last_link_header)
    }
  end

  def delete!(url, headers) do
    request = %{url: url, headers: headers}
    response = HTTPoison.delete!(url, headers)

    %Github.Client.Response{
      request: request,
      body: from_json!(response.body),
      headers: response.headers,
      status: response.status_code
    }
  end

  def patch!(url, headers, body) do
    request = %{url: url, headers: headers, body: body}
    response = HTTPoison.patch!(url, body, headers)

    %Github.Client.Response{
      request: request,
      body: from_json!(response.body),
      headers: response.headers,
      status: response.status_code
    }
  end

  @doc """
  The next paginated page

  ## Example

      iex> client = %Github.Client{access_token: "access_token"}
      iex> response1 = client |> Github.Apps.Installations.list_repos!(per_page: 1)
      %Github.Client.Response{
        next_url: "https://api.github.com/installation/repositories?per_page=1&page=2",
        last_url: "https://api.github.com/installation/repositories?per_page=1&page=77",
        status: 200,
        ...
      }

      iex> response2 = client.fetch_more!(response1)
      %Github.Client.Response{
        next_url: "https://api.github.com/installation/repositories?per_page=1&page=3",
        last_url: "https://api.github.com/installation/repositories?per_page=1&page=77",
        status: 200,
        ...
      }
  """
  def fetch_more!(github_response) do
    case github_response.next_url do
      nil -> nil
      url -> get!(url, github_response.request.headers)
    end
  end

  @doc """
  The rest of the paginated pages

  ## Example

      iex> client = %Github.Client{access_token: "access_token"}
      iex> response1 = client |> Github.Apps.Installations.list_repos!(per_page: 30)
      %Github.Client.Response{
        next_url: "https://api.github.com/installation/repositories?per_page=30&page=2",
        last_url: "https://api.github.com/installation/repositories?per_page=30&page=3",
        status: 200,
        ...
      }

      iex> responses = client.fetch_all!(response1)
      [
        %Github.Client.Response{
          next_url: "https://api.github.com/installation/repositories?per_page=30&page=3",
          last_url: "https://api.github.com/installation/repositories?per_page=30&page=3",
          status: 200,
          ...
        },
        %Github.Client.Response{
          next_url: nil,
          last_url: "https://api.github.com/installation/repositories?per_page=30&page=3",
          status: 200,
          ...
        }
      ]

  To reduce number of HTTP requests it is recommended to specify the largest `per_page` value `100` in the first request.
  """
  def fetch_all!(github_response) do
    github_response |> fetch_all!([]) |> Enum.reverse()
  end

  @doc """
  Generate GitHub [JWT token](https://developer.github.com/apps/building-github-apps/authenticating-with-github-apps/#jwt-payload) with App ID and private key

  ## Example

      iex> Github.Client.generate_jwt_token(id: 12345, private_key_filepath: "app.pem", expire_in_minutes: 10)
      "eyJhbGciOiJSU..."
  """
  def generate_jwt_token(options) do
    opts = Enum.into(options, @generate_jwt_token_default_options)
    pem = File.read!(opts.private_key_filepath)
    signer = Joken.Signer.create("RS256", %{"pem" => pem})

    %{
      default_exp: opts.expire_in_minutes * 60,
      iss: opts.app_id |> Integer.to_string()
    }
    |> Joken.Config.default_claims()
    |> Joken.generate_and_sign!(nil, signer)
  end

  defp json_library do
    Application.get_env(Github.Client, :json_library) || Jason
  end

  defp fetch_all!(github_response, acc) do
    github_response = fetch_more!(github_response)

    case github_response do
      nil -> acc
      _ -> fetch_all!(github_response, [github_response | acc])
    end
  end

  defp next_link_header(response) do
    link_header = Enum.find(response.headers, fn {header, _value} -> header == "Link" end)

    case link_header do
      nil ->
        nil

      {_, link_value} ->
        link_value |> String.split(", ") |> Enum.find(fn link -> link =~ "rel=\"next\"" end)
    end
  end

  defp last_link_header(response) do
    link_header = Enum.find(response.headers, fn {header, _value} -> header == "Link" end)

    case link_header do
      nil -> nil
      {_, link_value} -> link_value |> String.split(", ") |> Enum.find(&(&1 =~ "rel=\"last\""))
    end
  end

  defp link_url(link_header) do
    case link_header do
      nil -> nil
      _ -> @link_url_regex |> Regex.named_captures(link_header) |> Map.get("url")
    end
  end

  defp link_page(link_header) do
    case link_header do
      nil -> nil
      _ -> @link_page_regex |> Regex.named_captures(link_header) |> Map.get("page")
    end
  end
end
